### PR TITLE
explicitly raise error if trainable params integers

### LINKED ISSUE #4835: qml.jacobian returns 0 for all circuits
### Expected behavior

```python
def circuit_function(weights):
    qml.Hadamard(wires=0)
    qml.RY(weights[0], wires=0)
    return [ qml.expval(qml.PauliZ(wires=0)) ]

weights         = [0]
control_weights = np.array(weights, requires_grad=True)

dev   = qml.device("default.qubit", wires=1, shots=None)
qNode = qml.QNode(circuit_function, dev, diff_method="backprop")

gradients   = qml.jacobian(lambda x: np.stack(qNode(x)))(control_weights)
slope       = sum(gradients)
print(f'slope={slope}')
print(f'gradients={gradients}')
```

-------------------------------
Expected behavour is for gradients to return

```python
slope=[-1.]
gradients=[[-1.]]
```

### Actual behavior

```python
slope=[0]
gradients=[[0]]
```

change weights to [0.0] fixed this problem

```python
slope=[-1.]
gradients=[[-1.]]
```

### Additional information

This behaviour was really hard to track down and could easily break code.  PennyLane code should be assuming that rotations are floating point values even if presented in the control weights as integer values.

Thanks,
Ian
cs.uwaterloo.ca/~ijdavis

### Source code

```shell
def circuit_function(weights):
    qml.Hadamard(wires=0)
    qml.RY(weights[0], wires=0)
    return [ qml.expval(qml.PauliZ(wires=0)) ]

weights         = [0]
control_weights = np.array(weights, requires_grad=True)

dev   = qml.device("default.qubit", wires=1, shots=None)
qNode = qml.QNode(circuit_function, dev, diff_method="backprop")

gradients   = qml.jacobian(lambda x: np.stack(qNode(x)))(control_weights)
slope       = sum(gradients)
print(f'slope={slope}')
print(f'gradients={gradients}')
```

### Tracebacks

_No response_

### System information

```python
Name: PennyLane
Version: 0.33.1
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane-Lightning

Platform info:           Linux-5.15.120+-x86_64-with-glibc2.35
Python version:          3.10.12
Numpy version:           1.23.5
Scipy version:           1.11.3
Installed devices:
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-50174] Fixes #4835 

Autograd can give confusing null behaviour when the "trainable" parameters are integers. This PR explicitly raises errors if any of the trainable parameters are integers.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -55,6 +55,8 @@
 
 <h3>Deprecations 👋</h3>
 
+* `qml.grad` and `qml.jacobian` now explicitly raise errors if trainable parameters are integers.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:58]:
```suggestion
* `qml.grad` and `qml.jacobian` now explicitly raise errors if trainable parameters are integers.
  [(#4836)](https://github.com/PennyLaneAI/pennylane/pull/4836)
```

**Review by lillian542 — APPROVED** [review]

File: pennylane/_grad.py [hunk 2]
```diff
@@ -135,7 +137,7 @@ def _grad_with_forward(fun, x):
         value."""
         vjp, ans = _make_vjp(fun, x)
 
-        if not vspace(ans).size == 1:
+        if vspace(ans).size != 1:
```
Comment (trbromley) [inline pennylane/_grad.py:140]: Is this just a tidier way of doing things?

Comment (albi3ro) [reply pennylane/_grad.py:140]: yep.  my IDE was complaining.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -55,6 +55,9 @@
...
+  [(#4836)](https://github.com/PennyLaneAI/pennylane/pull/4836)
+
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:60]: Would this be more of an improvement or bugfix than a deprecation?

Comment (albi3ro) [reply doc/releases/changelog-dev.md:60]: Eh... oops?

Comment (trbromley) [reply doc/releases/changelog-dev.md:60]: Fixed :+1:

**Comment by trbromley** [comment]
I'm thinking we might want to revisit this PR.

For example, now in v0.34 it is not possible to do:
```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit")

@qml.qnode(dev)
def f(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

x = np.array(1, requires_grad=True)

qml.grad(f)(x)
```
In v0.33 and earlier, this was correctly calculated to be `-0.84147098`. It's a little annoying to have to put `1.`, as @alvaro-at-xanadu can attest to while having to write challenges for QHack.

Could an alternative approach be that we cast to floats if `requires_grad=True`?

**Comment by josh146** [comment]
@trbromley if it behaved differently in versions 0.33 and lower, and now raises an error, I would say this is a breaking change.

note that this will likely also affect existing coding challenges and education materials/demos, so recommend we keep the previous behaviour

**Comment by albi3ro** [comment]
I would compare this to:
```
import autograd

def f(x):
    return x**2

autograd.grad(f)(autograd.numpy.array(1))
```
which raises:
```
TypeError: Can't differentiate w.r.t. type <class 'numpy.int64'>
```

So we are simply matching autograd behaviour in our wrapped version of autograd.

**Pull request merged**